python/cudf/cudf/io/parquet.py

```diff
@@ -790,7 +780,199 @@ def merge_parquet_filemetadata(filemetadata_list):
+
+        (
+            paths,
+            metadata_file_paths,
+            grouped_df,
+            offsets,
+            self.filename,
+        ) = _get_partitioned(
+            df,
+            self.path,
+            self.partition_cols,
+            preserve_index=self.common_args["index"],
+            filename=self.filename,
+        )
+
+        existing_cw_batch = defaultdict(dict)
+        new_cw_paths = []
+
+        def pairwise(iterable):
+            it = iter(iterable)
+            a = next(it, None)
+            for b in it:
+                yield (a, b - a)
+                a = b
+
+        for path, part_info, meta_path in zip(
+            paths, pairwise(offsets), metadata_file_paths
+        ):
+            if path in self.path_cw_map:  # path is a currently open file
+                cw_idx = self.path_cw_map[path]
+                existing_cw_batch[cw_idx][path] = part_info
+            else:  # path not currently handled by any chunked writer
+                new_cw_paths.append((path, part_info, meta_path))
+
+        # Write out the parts of grouped_df currently handled by existing cw's
+        for batch_cw_path_list in existing_cw_batch.items():
+            cw_idx = batch_cw_path_list[0]
+            cw = self._chunked_writers[cw_idx][0]
+            # match found paths with this cw's paths and nullify partition info
+            # for partition_col values not in this batch
+            this_cw_part_info = [
+                batch_cw_path_list[1].get(path, (0, 0))
+                for path in self._chunked_writers[cw_idx][1]
+            ]
+            cw.write_table(grouped_df, this_cw_part_info)
+
+        # Create new cw for unhandled paths encountered in this write_table
+        new_paths = [path for path, _, _ in new_cw_paths]
+        meta_paths = [path for _, _, path in new_cw_paths]
+        self._chunked_writers.append(
+            (
+                libparquet.ParquetWriter(new_paths, **self.common_args),
+                new_paths,
+                meta_paths,
+            )
+        )
+        new_cw_idx = len(self._chunked_writers) - 1
+        self.path_cw_map.update({k: new_cw_idx for k in new_paths})
+        part_info = [info for _, info, _ in new_cw_paths]
+        self._chunked_writers[-1][0].write_table(grouped_df, part_info)
+
+    def close(self, metadata_file_path=None):
+        """
+        Close all open files and optionally return footer metadata as a binary
+        blob
+        """
+        return_metadata = bool(metadata_file_path)
+        if self.partition_cols is not None:
+            if isinstance(metadata_file_path, str):
+                warnings.warn(
+                    "metadata_file_path is automatically determined for "
+                    "partitioned writing. The passed metadata_file_path will "
+                    "be ignored"
+                )
+        else:
+            if return_metadata:
+                cw, paths, _ = self._chunked_writers.pop()
```

Comment:
Since tuples are immutable, when we wanted to add a field `metadata_file_path` which was empty in the ctor, I popped and added it back. The want to confirm that the first object `cw` is never copied. It's either moved or that `cw` acts as a reference/shared pointer.

Reply:
I think this is fine.